Libc/string/memset.3
.\" Copyright (c) 1990, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Chris Torek and the American National Standards Committee X3,
.\" on Information Processing Systems.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)memset.3	8.1 (Berkeley) 6/4/93
.\" $FreeBSD: src/lib/libc/string/memset.3,v 1.7 2001/10/01 16:09:00 ru Exp $
.\"
.Dd June 4, 1993
.Dt MEMSET 3
.Os
.Sh NAME
.Nm memset
.Nd fill a byte string with a byte value
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In string.h
.Ft void *
.Fn memset "void *b" "int c" "size_t len"
.Sh DESCRIPTION
The
.Fn memset
function
writes
.Fa len
bytes of value
.Fa c
(converted to an unsigned char) to the byte string
.Fa b .
.Sh RETURN VALUES
The
.Fn memset
function returns its first argument.
.Sh SEE ALSO
.Xr bzero 3 ,
.Xr memset_pattern 3 ,
.Xr swab 3
.Sh STANDARDS
The
.Fn memset
function
conforms to
.St -isoC .


X11/ChangeLog
2006-09-08  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_44

2006-09-06  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.h,
	xc/lib/GL/apple/dri_dispatch.defs: removed glPointParameteri[v]
	aliases

	* xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/applewmstr.h,
	xc/programs/Xserver/hw/apple/applewm.h,
	xc/programs/Xserver/hw/apple/applewm.c,
	xc/programs/Xserver/hw/apple/applewm-impl.h,
	xc/programs/Xserver/hw/apple/applewm-impl.c,
	xc/programs/Xserver/hw/apple/X11Controller.m,
	xc/programs/Xserver/hw/apple/X11Application.m: use shared
	version of AppleWM headers

	* make-host-def, build.sh, Makefile: removed Xplugin/quartz-wm
	references; attempt to support x86_64

2006-07-07  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_43

	* xc/programs/Xserver/hw/apple/x-hash.c: fix compiler warning

	* xc/programs/xterm/main.c: fix xterm build failure (4607761)

2006-02-12  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/GL/apple/indirect.c
	(glAquaInitVisualConfigs): add stereo visuals (4377346)

2006-01-16  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_42

	* xc/programs/Xserver/hw/apple/bundle-main.c: added /sbin and
	/usr/sbin to the default PATH value (4408133)

	* xc/programs/Xserver/hw/apple/quartz.c (do_exec): use same
	default PATH and MANPATH values as bundle-main.c (4408133)

2005-11-29  John Harper  <jsh@apple.com>

	Fix xscreensaver-demo crash and make its cross-process GL
	surfaces work correctly (4356702):

	* xc/programs/Xserver/hw/apple/dri-surface.c: don't compile the
	code to draw on surfaces (that was always disables but was
	causing a crash - I've fixed the crash, but the code is never
	used so best to just remove it from the binary)

2005-10-25  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_41

	* xc/lib/xtrans/Xtranssock.c: fix Leopard build failure
	(4222517)

2005-10-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: fix default
	MANPATH (4312329)

2005-08-29  John Harper  <jsh@apple.com>

	* xc/extras/freetype2/src/base/ftmac.c: merged endian changes
	from 3477131

2005-06-22  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: merged
	fix for CAN-2005-0605 from xfree86 cvs (4031381)

2005-05-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c
	(install_ipaddr_source): changed the way we listen for network
	/ hostname changes, hopefully to be more reliable

	* xc/programs/Xserver/hw/apple/X11Application.m (sendEvent):
	change fix for bring-all-to-front not to require private
	headers - hardcode enum value we need :-(  (4075093)

2005-05-23  John Harper  <jsh@apple.com>

	* Makefile, make-host-def, merge-libs.sh,
	xc/config/cf/Imake.cf, xc/config/cf/darwin.cf,
	xc/config/cf/darwinLib.tmpl, xc/include/Xmd.h,
	xc/lib/Xt/TMaction.c, xc/lib/Xt/Xtos.h,
	xc/lib/freetype2/Imakefile: checked in changes to
	support building fat libraries when ppc64 is in RC_ARCHS
	(3721701)

	* xc/programs/Xserver/hw/apple/X11.pbproj/project.pbxproj:
	bumped short version number to 1.2 so we don't fail to verify

	* xc/programs/Xserver/hw/apple/X11Application.m: handle scroll
	wheel events with delta < 1 correctly (4109092)

2005-05-20  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/rootless.h,
	xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Application.m: recognize the
	bfCPSOrderAllWindowsForward bit in activation events and
	respond to it appropriately (4075093)

2005-03-09  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_40

	* xc/programs/Xserver/hw/apple/darwin.c
	(DarwinUpdateModifiers): handle AlphaShift (caps lock) as a
	normal modifier, now that we disabled XKB (when merging with
	XFree86 4.4) that's the correct thing to do (4015381)

2005-03-03  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_39

2005-03-01  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: more
	Xpm fixes, from
	https://bugs.freedesktop.org/show_bug.cgi?id=1920, attachment
	1909. (4031381)

	* xc/lib/Xpm/Imakefile: doh, forgot Imakefile portion of
	original patch

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/s_popen.c,
	xc/extras/Xpm/lib/parse.c, xc/extras/Xpm/lib/misc.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/WrFFrBuf.c,
	xc/extras/Xpm/lib/RdFToI.c, xc/extras/Xpm/lib/RdFToBuf.c,
	xc/extras/Xpm/lib/Imakefile, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/CrBufFrI.c, xc/extras/Xpm/lib/Attrib.c:
	merged changes from XFree86 HEAD branch (4022757)

2005-02-15  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_38

	* xc/programs/Xserver/hw/apple/quartz-server.c: request that
	deferred updates are disabled. Otherwise we lose 25%
	performance on the "cat file in xterm" test (3985081)

	Fix 3494176 (deadlock while closing OpenGL window):

	* xc/lib/GL/glx/glxext.c: make __glXLock a recursive mutex

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	make the screen mutex a pthread_mutex and make it recursive

2005-01-04  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_37

	* xc/programs/Xserver/hw/apple/X11Controller.m: fixed several
	bugs in the customization dialog for the applications menu
	(3668769)

2004-11-24  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36_1

	* Makefile: strip-libraries target doesn't strip libXplugin

2004-11-23  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36

2004-11-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: added a signal
	handler to catch SIGTERM etc and exit cleanly, killing the X
	server and xinit (3879878)


Libc/ppc/string/memset.s
/*
 * Copyright (c) 2000-2004 Apple Computer, Inc. All rights reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 * 
 * This file contains Original Code and/or Modifications of Original Code
 * as defined in and that are subject to the Apple Public Source License
 * Version 2.0 (the 'License'). You may not use this file except in
 * compliance with the License. Please obtain a copy of the License at
 * http://www.opensource.apple.com/apsl/ and read it before using this
 * file.
 * 
 * The Original Code and all software distributed under the License are
 * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
 * Please see the License for the specific language governing rights and
 * limitations under the License.
 * 
 * @APPLE_LICENSE_HEADER_END@
 */

#include <machine/cpu_capabilities.h>

/* We use mode-independent "g" opcodes such as "srgi".  These expand
 * into word operations when targeting __ppc__, and into doubleword
 * operations when targeting __ppc64__.
 */
#include <architecture/ppc/mode_independent_asm.h>


        .text
#define kShort  128             // threshold for calling commpage


/* ***************
 * * M E M S E T *
 * ***************
 *
 * Registers we use:
 *      r3  = original ptr, not changed since memset returns it
 *      r4  = count of bytes to set
 *      r7  = value to set
 *      r8  = working operand ptr
 */
 
        .globl	_memset
        .align	5
_memset:                        // void *   memset(void *b, int c, size_t len);
        andi.	r7,r4,0xFF      // copy value to working register, test for 0
        mr	r4,r5           // move length to working register
        cmplgi	cr1,r5,kShort	// long enough to bother with _COMM_PAGE_MEMSET_PATTERN?
        beqa++	_COMM_PAGE_BZERO    // if (c==0), map to bzero()
        rlwimi	r7,r7,8,16,23	// replicate nonzero value to low 2 bytes
        neg	r5,r3           // start to compute #bytes to align
        mr	r8,r3           // make working copy of operand ptr
        rlwimi	r7,r7,16,0,15	// value now in all 4 bytes
        blt	cr1,Lmemset3    // too short to use commpage
        andi.	r0,r5,0xF       // r0 <- #bytes to align on quadword
        
        // Align ptr and store enough so that we have an aligned 16-byte pattern.

        stw     r7,0(r8)
        stw     r7,4(r8)
        stw     r7,8(r8)
        stw     r7,12(r8)
        beq     Lmemset1        // skip if (r0==0), ie if r8 is 16-byte aligned
        add     r8,r8,r0        // 16-byte align ptr
        sub     r4,r4,r0        // adjust length
        stw     r7,0(r8)        // now we can store an aligned 16-byte pattern
        stw     r7,4(r8)
        stw     r7,8(r8)
        stw     r7,12(r8)

        // Call machine-specific commpage routine, which expects:
        //      r4 = count (>=32)
        //      r8 = ptr (16-byte aligned) to memory to store
        //      r9 = ptr (16-byte aligned) to 16-byte pattern to store
        // When it returns:
        //      r3, r7, and r12 are preserved
        //      r4 and r8 are updated to reflect a residual count of from 0..31 bytes
        
Lmemset1:
        mflr    r12             // save return address
        mr      r9,r8           // point to 16-byte-aligned 16-byte pattern
        addi    r8,r8,16        // point to first unstored byte
        subi    r4,r4,16        // account for the aligned bytes we have stored
        bla	_COMM_PAGE_MEMSET_PATTERN
        mtlr    r12

        // Here for short nonzero memset.
        //  r4 = count (<= kShort bytes)
        //  r7 = pattern in all four bytes
        //  r8 = ptr
Lmemset3:
        srgi.   r0,r4,4         // any 16-byte chunks?
        mtcrf   0x01,r4         // move length remaining to cr7 so we can test bits
        beq     Lmemset5        // fewer than 16 bytes
        mtctr   r0
        b       Lmemset4        // enter loop
        
        .align  5
Lmemset4:                       // loop over 16-byte chunks
        stw     r7,0(r8)
        stw     r7,4(r8)
        stw     r7,8(r8)
        stw     r7,12(r8)
        addi    r8,r8,16
        bdnz++  Lmemset4
        
        // Handle last 0..15 bytes.
Lmemset5:
        bf      28,2f
        stw     r7,0(r8)
        stw     r7,4(r8)
        addi    r8,r8,8
2:
        bf      29,3f
        stw     r7,0(r8)
        addi    r8,r8,4
3:
        bf      30,4f
        sth     r7,0(r8)
        addi    r8,r8,2
4:
        bflr    31
        stb     r7,0(r8)
        blr
        

/* ***********************************
 * * M E M S E T _ P A T T E R N 1 6 *
 * ***********************************
 *
 * Used to store a 16-byte pattern in memory:
 *
 *  void    memset_pattern16(void *b, const void *c16, size_t len);
 *
 * Where c16 points to the 16-byte pattern.  None of the parameters need be aligned.
 */

        .globl	_memset_pattern16
        .align	5
_memset_pattern16:
        cmplgi  cr1,r5,kShort   // check length
        lwz     r7,0(r4)        // load pattern into (these remain lwz in 64-bit mode)
        lwz     r9,4(r4)
        neg     r6,r3           // start to compute ptr alignment
        lwz     r10,8(r4)
        lwz     r11,12(r4)
        b       __memset_pattern_common
        

/* *********************************
 * * M E M S E T _ P A T T E R N 8 *
 * *********************************
 *
 * Used to store an 8-byte pattern in memory:
 *
 *  void    memset_pattern8(void *b, const void *c8, size_t len);
 *
 * Where c8 points to the 8-byte pattern.  None of the parameters need be aligned.
 */

        .globl	_memset_pattern8
        .align	5
_memset_pattern8:
        lwz     r7,0(r4)        // load pattern (these remain lwz in 64-bit mode)
        lwz     r9,4(r4)
        cmplgi  cr1,r5,kShort   // check length
        neg     r6,r3           // start to compute ptr alignment
        mr      r10,r7          // replicate into 16-byte pattern
        mr      r11,r9
        b       __memset_pattern_common
        

/* *********************************
 * * M E M S E T _ P A T T E R N 4 *
 * *********************************
 *
 * Used to store a 4-byte pattern in memory:
 *
 *  void    memset_pattern4(void *b, const void *c4, size_t len);
 *
 * Where c4 points to the 4-byte pattern.  None of the parameters need be aligned.
 */

        .globl	_memset_pattern4
        .align	5
_memset_pattern4:
        lwz     r7,0(r4)        // load pattern
        cmplgi  cr1,r5,kShort   // check length
        neg     r6,r3           // start to compute ptr alignment
        mr      r9,r7           // replicate into 16-byte pattern
        mr      r10,r7
        mr      r11,r7
        b       __memset_pattern_common // don't fall through because of scatter-loading
        
        
/* ***********************************************
 * * _ M E M S E T _ P A T T E R N _ C O M M O N *
 * ***********************************************
 *
 * This is the common code used by _memset_pattern16, 8, and 4.  They all get here via
 * long branch (ie, "b") in case the routines are re-ordered, with:
 *      r3 = ptr to memory to store pattern into (unaligned)
 *      r5 = length in bytes
 *      r6 = neg(r3), used to compute #bytes to align
 *      r7, r9, r10, r11 = 16-byte pattern to store
 *      cr1= ble if (r5 <= kShort)
 */

        .globl	__memset_pattern_common
        .private_extern __memset_pattern_common // avoid dyld stub, which trashes r11
        .align	5
__memset_pattern_common:
        andi.   r0,r6,0xF       // get #bytes to 16-byte align ptr
        ble--   cr1,LShort      // if short operand skip out

        // Align ptr and store enough of pattern so we have an aligned
        // 16-byte chunk of it (this effectively rotates incoming pattern
        // if the original ptr was not aligned.)
        
        stw     r7,0(r3)
        stw     r9,4(r3)
        stw     r10,8(r3)
        stw     r11,12(r3)
        beq     Laligned        // skip if (r0==0), ie if r3 is 16-byte aligned
        stw     r7,16(r3)
        stw     r9,20(r3)
        stw     r10,24(r3)
        stw     r11,28(r3)
        add     r3,r3,r0        // 16-byte align ptr
        sub     r5,r5,r0        // adjust length
        
        // We're ready to call the machine-specific commpage routine
        // to do the heavy lifting.  When called, _COMM_PAGE_MEMSET_PATTERN expects:
        //      r4 = length (>= 32)
        //      r8 = ptr (16-byte aligned)
        //      r9 = ptr to 16-byte pattern (16-byte aligned)
        // When it returns:
        //      r3, r7, and r12 are preserved
        //      r4 and r8 are updated to reflect a residual count of from 0..31 bytes

Laligned:
        mflr    r12             // save return across commpage call
        mr      r9,r3           // point to 16-byte aligned 16-byte pattern
        addi    r8,r3,16        // point to first unstored byte (r8 is 16-byte aligned)
        subi    r4,r5,16        // account for the aligned bytes we have stored
        bla     _COMM_PAGE_MEMSET_PATTERN
        mr.     r5,r4           // move length (0..31) back to original reg and test for 0
        mtlr    r12
        beqlr                   // done if residual length == 0
        lwz     r7,-16(r8)      // load aligned pattern into r7,r9,r10, and r11
        lwz     r9,-12(r8)
        mr      r3,r8           // move destination ptr back
        lwz     r10,-8(r8)
        lwz     r11,-4(r8)
        
        // Handle short operands and leftovers.
        //      r3 = dest
        //      r5 = length
        //      r7,r9,r10,r11 = pattern
LShort:
        srgi.   r0,r5,4         // at least 16 bytes?
        mtcrf   0x01,r5         // move leftover count to cr7
        beq     Lleftovers
        mtctr   r0
LShortLoop:
        stw     r7,0(r3)        // replicate the pattern
        stw     r9,4(r3)
        stw     r10,8(r3)
        stw     r11,12(r3)
        addi    r3,r3,16
        bdnz    LShortLoop      // store 16 more bytes
        
        // Fewer than 16 bytes remaining.
Lleftovers:        
        bf      28,1f
        stw     r7,0(r3)        // store next 8 bytes
        stw     r9,4(r3)
        addi    r3,r3,8
        mr      r7,r10          // shift pattern over
        mr      r9,r11
1:
        bf      29,2f
        stw     r7,0(r3)
        addi    r3,r3,4
        mr      r7,r9
2:
        bf      30,3f
        rlwinm  r7,r7,16,0,31   // position leftmost 2 bytes for store
        sth     r7,0(r3)
        addi    r3,r3,2
3:
        bflr    31
        srwi    r7,r7,24        // position leftmost byte for store
        stb     r7,0(r3)
        blr
